### PR TITLE
feat(tests): add enotdir tests

### PR DESCRIPTION
**Pull request by musikid**
This is the first PR related to the error tests.
It adds support for the ENOTDIR tests, by providing a macro which can be used in the syscalls directory.

### CODE REVIEW

File: rust/src/tests/errors/enotdir.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        fn enotdir_component(ctx: &mut crate::runner::context::TestContext,
+                            ft: crate::runner::context::FileType) {
+            let base_path = ctx.create(ft.clone()).unwrap();
+            let path = base_path.join("previous_not_dir");
+
+            assert_eq!($f(ctx, &path).unwrap_err(), nix::errno::Errno::ENOTDIR)
+        }
+    };
+
+    ($syscall: ident $( ($( $($before:expr),* ,)? ~path $(, $($after:expr),*)?) )?) => {
```
Comment (asomers) [inline rust/src/tests/errors/enotdir.rs:27]: I find this syntax to be pretty weird, and it doesn't seem to save much code.  I think it would be clearer to remove this macro form, and write
```rust
assert_enotdir_comp!(mkdir, |_ctx, path| mkdir(path, Mode::empty()));
```
instead of
```rust
assert_enotdir_comp!(mkdir(~path, Mode::empty()));
```

File: rust/src/tests/errors/enotdir.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/// Asserts that it returns ENOTDIR if a component of the path prefix is not a directory.
+macro_rules! assert_enotdir_comp {
```
Comment (asomers) [inline rust/src/tests/errors/enotdir.rs:2]: The name of this macro suggests that it expands to a statement, something similar to `assert!`.  But it actually expands to an item.  I suggest renaming it to something like `enotdir_test_case!`.

File: rust/src/tests/mknod.rs [hunk 3]
```diff
@@ -49,3 +51,25 @@ fn changed_time_fields_success(ctx: &mut TestContext) {
             mknod_wrapper(&path, Mode::from_bits_truncate(0o644)).unwrap();
         });
 }
+
+// mknod/01.t
+assert_enotdir_comp!(mknod(~path, SFlag::S_IFIFO, Mode::empty(), 0));
+
+crate::test_case! {
```
Comment (asomers) [inline rust/src/tests/mknod.rs:58]: Is the only reason that this can't use the macro the fact that it needs root?

File: rust/src/tests/chmod.rs [hunk 4]
```diff
@@ -101,3 +107,77 @@ fn clear_isgid_bit(ctx: &mut SerializedTestContext) {
     assert_eq!(actual_mode & 0o7777, expected_mode.bits());
     //TODO: FreeBSD "S_ISGID should be removed and chmod(2) should success and FreeBSD returns EPERM."
 }
+
+crate::test_case! {
```
Comment (asomers) [inline rust/src/tests/chmod.rs:111]: This sticky stuff doesn't have anything to do with ENOTDIR, right?  It should go in a separate PR.

**Review by asomers — CHANGES_REQUESTED** [review]

File: rust/src/tests/errors/enotdir.rs (hunk 1 above)
Comment (asomers) [reply rust/src/tests/errors/enotdir.rs:27]: After spending some more time reading your PRs, I'm getting more comfortable with this unique syntax.  I think it's ok to leave it here, but please add some documentation to the macro explaining what it means.  Doc tests would be useful too.

File: rust/src/tests/chmod.rs (hunk 4 above)
Comment (musikid) [reply rust/src/tests/chmod.rs]: Yes, sorry!

File: rust/src/tests/mknod.rs (hunk 3 above)
Comment (musikid) [reply rust/src/tests/mknod.rs:58]: No, should I add a macro form for this?

Comment (asomers) [reply rust/src/tests/mknod.rs:58]: What else does it need?

File: rust/src/tests/errors/enotdir.rs [hunk 5]
```diff
@@ -1,15 +1,25 @@
...
-                let file = path.parent().unwrap();
-                $syscall(path, &new_path).or($syscall(file, path))
-        });
-    };
-
-    ($syscall: ident, $f: expr) => {
+/// Create a test case which asserts that the syscall returns ENOTDIR
+/// if a component of the path prefix is not a directory.
+/// There are multiple forms for this macro:
+/// - A basic form which takes the syscall and optionally a `~path` argument if the path
```
Comment (asomers) [inline rust/src/tests/errors/enotdir.rs:4]: Yeah, but what does `~path` mean, and where should one put it?

File: rust/src/tests/mknod.rs (hunk 3 above)
Comment (musikid) [reply rust/src/tests/mknod.rs:58]: Sorry, it doesn't need anything besides the root privileges.

Comment (asomers) [reply rust/src/tests/mknod.rs:58]: Well, it's up to you.  If you can save code by moving some logic into the macro, then I suggest you do.  But if this is the only place where you would need it, then it might not be worthwhile.

**Review by asomers — APPROVED** [review]

File: rust/src/tests/chown.rs [hunk 6]
```diff
@@ -2,17 +2,17 @@ use nix::unistd::chown;
...
+enotdir_comp_test_case!(chown, |ctx: &mut TestContext, path| {
     let user = ctx.get_new_user();
     chown(path, Some(user.uid), None)
 });
 
 mod lchown {
     use super::*;
 
-    assert_enotdir_comp!(lchown, |ctx: &mut TestContext, path| {
+    enotdir_comp_test_case!(lchown, |ctx: &mut TestContext, path| {
```
Comment (musikid) [inline rust/src/tests/chown.rs:15]: I'm not sure if it's better to separate the lchown/lchmod in other modules since we test them in the same test functions before. We could make the macro take multiple functions to test them in the same test function. What do you think about it?

Comment (asomers) [reply rust/src/tests/chown.rs:15]: Better to leave them separate.  Combining them into the same function, as you do in PR #108 , isn't great.  In that case I think it's OK for now because tests/chown/00.t is so complicated that it would be hard to separate them at the same time that you convert them.  Ideally, we would separate them later.  But here, separation is better, as usual.
